terminus-ui/drawer/src/drawer.component.ts

```diff
@@ -0,0 +1,343 @@
+  }
+
+  // Event emitted when the drawer has started collapsing.
+  @Output()
+  public get collapsedStart(): Observable<void> {
+    return this.animationStarted.pipe(
+      filter(e => e.fromState !== e.toState && e.toState === 'void'),
+      untilComponentDestroyed(this),
+      map(() => {})
+    );
+  }
+
+  // Event emitted when the drawer's position changes.
+  // tslint:disable-next-line:no-output-on-prefix
+  @Output('positionChanged')
+  public readonly onPositionChanged: EventEmitter<void> = new EventEmitter<void>();
+
+  constructor(public elementRef: ElementRef<HTMLElement>,
+              private platform: Platform,
+              private ngZone: NgZone,
+              public renderer: Renderer2,
+              // tslint:disable-next-line:no-any
+              @Optional() @Inject(DOCUMENT) private doc: any) {
+
+    /**
+     * Listen to `keydown` events outside the zone so that change detection is not run every
+     * time a key is pressed. Re-enter the zone only if the `ESC` key is pressed
+     * and we don't have close disabled.
+     */
+    this.ngZone.runOutsideAngular(() => {
+      (fromEvent(this.elementRef.nativeElement, 'keydown') as Observable<KeyboardEvent>)
+        .pipe(
+          filter(event => event.code === KEYS.ESCAPE.code && !hasModifierKey(event)),
+          untilComponentDestroyed(this)
+        ).subscribe(event => this.ngZone.run(() => {
+          this.collapse();
+          event.stopPropagation();
+          event.preventDefault();
+        }));
+    });
+
+    // We need a Subject with distinctUntilChanged, because the `done` event
+    // fires twice on some browsers.
+    this.animationEnd.pipe(
+      distinctUntilChanged((x, y) => x.fromState === y.fromState && x.toState === y.toState),
+      untilComponentDestroyed(this),
+    ).subscribe((event: AnimationEvent) => {
+      const { fromState, toState } = event;
+
+      if ((toState.indexOf('open') === 0 && fromState === 'void')
+        || (toState === 'void' && fromState.indexOf('open') === 0)
+      ) {
+        this.expandedChange.emit(this.isExpanded);
+      }
+    });
+    this.renderer.setStyle(this.elementRef.nativeElement, 'width', this._expandedSize);
+  }
+
+  public ngAfterContentChecked() {
+    // Enable the animations after the lifecycle hooks have run, in order to avoid animating
+    // drawers that are open by default.
+    if (this.platform.isBrowser) {
+      this.enableAnimations = true;
+    }
+  }
+
+  public ngOnDestroy() {
+    this.modeChanged.complete();
+    this.destroyed.next();
+    this.destroyed.complete();
+  }
+
+  /**
+   * Expand the drawer.
+   *
+   * @return Promise<TsDrawerToggleResult>
+   */
+  public expand(): Promise<TsDrawerToggleResult> {
+    return this.toggle(true);
+  }
+
+  /**
+   * Collapse the drawer.
+   *
+   * @return Promise<TsDrawerToggleResult>
+   */
+  public collapse(): Promise<TsDrawerToggleResult> {
+    this.elementRef.nativeElement.classList.add('ts-drawer--collapsed');
```

Comment:
Seems like class addition/removal would be better suited as part of the `toggle` method?

Reply:
I guess it's fine either way... had preferred here because it was more clear here versus in toggle, it'd live in a condition block.

Reply:
I only say that because this function is not actually in control of collapsing the drawer. It just calls another function that is actually in charge of adjusting the drawer.